modules/layers/src/text-layer/font-atlas-manager.js

```diff
@@ -0,0 +1,264 @@
+export const DEFAULT_FONT_FAMILY = 'Monaco, monospace';
+export const DEFAULT_FONT_WEIGHT = 'normal';
+export const DEFAULT_FONT_SIZE = 64;
+export const DEFAULT_BUFFER = 2;
+export const DEFAULT_CUTOFF = 0.25;
+export const DEFAULT_RADIUS = 3;
+
+const GL_TEXTURE_WRAP_S = 0x2802;
+const GL_TEXTURE_WRAP_T = 0x2803;
+const GL_CLAMP_TO_EDGE = 0x812f;
+const MAX_CANVAS_WIDTH = 1024;
+
+const BASELINE_SCALE = 0.9;
+const HEIGHT_SCALE = 1.2;
+
+// only preserve latest three fontAtlas
+const CACHE_LIMIT = 3;
+
+/**
+ * [key]: {
+ *   xOffset, // x position of last character in mapping
+ *   yOffset, // y position of last character in mapping
+ *   mapping, // x, y coordinate of each character in shared `fontAtlas`
+ *   data, // canvas
+ *   width. // canvas.width,
+ *   height, // canvas.height
+ * }
+ *
+ */
+const cache = new LRUCache(CACHE_LIMIT);
+
+const VALID_PROPS = [
+  'fontFamily',
+  'fontWeight',
+  'characterSet',
+  'fontSize',
+  'sdf',
+  'buffer',
+  'cutoff',
+  'radius'
+];
+
+/**
+ * get all the chars not in cache
+ * @param key cache key
+ * @param characterSet (Array|Set)
+ * @returns {Array} chars not in cache
+ */
+function getNewChars(key, characterSet) {
+  const cachedFontAtlas = cache.get(key);
+  if (!cachedFontAtlas) {
+    return characterSet;
+  }
+
+  const newChars = [];
+  const cachedMapping = cachedFontAtlas.mapping;
+  let cachedCharSet = Object.keys(cachedMapping);
+  cachedCharSet = new Set(cachedCharSet);
+
+  let charSet = characterSet;
+  if (charSet instanceof Array) {
+    charSet = new Set(charSet);
+  }
+
+  charSet.forEach(char => {
+    if (!cachedCharSet.has(char)) {
+      newChars.push(char);
+    }
+  });
+
+  return newChars;
+}
+
+function populateAlphaChannel(alphaChannel, imageData) {
+  // populate distance value from tinySDF to image alpha channel
+  for (let i = 0; i < alphaChannel.length; i++) {
+    imageData.data[4 * i + 3] = alphaChannel[i];
+  }
+}
+
+function setTextStyle(ctx, fontFamily, fontSize, fontWeight) {
+  ctx.font = `${fontWeight} ${fontSize}px ${fontFamily}`;
+  ctx.fillStyle = '#000';
+  ctx.textBaseline = 'baseline';
+  ctx.textAlign = 'left';
+}
+
+export default class FontAtlasManager {
+  constructor(gl) {
+    this.gl = gl;
+    this.fontFamily = DEFAULT_FONT_FAMILY;
+    this.fontWeight = DEFAULT_FONT_WEIGHT;
+    this.characterSet = DEFAULT_CHAR_SET;
+    this.fontSize = DEFAULT_FONT_SIZE;
+    this.buffer = DEFAULT_BUFFER;
+    // sdf only props
+    // https://github.com/mapbox/tiny-sdf
+    this.sdf = false;
+    this.cutoff = DEFAULT_CUTOFF;
+    this.radius = DEFAULT_RADIUS;
+
+    // key is used for caching generated fontAtlas
+    this._key = null;
+    this._texture = new Texture2D(this.gl);
+  }
+
+  get texture() {
+    return this._texture;
+  }
+
+  get mapping() {
+    const data = cache.get(this._key);
+    return data && data.mapping;
+  }
+
+  get scale() {
+    return HEIGHT_SCALE;
+  }
+
+  setProps(props = {}) {
```

Comment:
new func.

Also the main logic to decide if needs generating a new fontAtlas or if fontAtlas needs update